Corvus.Json.SourceGenerator generate Nullable types resulting in a compiler error.

Corvus.Json.SourceGenerator generates Nullable types resulting in a compiler error [CS0453](https://learn.microsoft.com/en-us/dotnet/csharp/misc/cs0453).

- schema
```json
{
    "$schema": "https://json-schema.org/draft-07/schema",
    "type": "object",
    "dependencies": {
        "recognition": { }
    }
}
```
- error
```
Restore complete (0.5s)
  Repro492 failed with 1 error(s) (0.8s)
    _\Corvus.Json.SourceGenerator\Corvus.Json.SourceGenerator.IncrementalSourceGenerator\PipelineSchema.Object.g.cs(695,30): error CS0453: The type 'JsonAny?' must be a non-nullable value type in order to use it as parameter 'T' in the generic type or method 'JsonValueExtensions.IsNotUndefined<T>(T)'
```

This is because `static void SomeExtensionMethods<T>(this T value) where T : struct`, is used directly for nullable types generated by the source generator.

The two available solutions are as follows:
- Not generating nullable types, similarly to the behaviour in Corvus.Json.JsonSchema.TypeGeneratorTool.
- Overload methods `static void OverloadExtensionMethods<T>(this Nullable<T> value) where T : struct`

This is the Minimum Repro Project: [Repro492.zip](https://github.com/user-attachments/files/17808389/Repro492.zip)

BTW: The two full schema I used are here: [pipeline.schema.json](https://raw.githubusercontent.com/MaaXYZ/MaaFramework/refs/heads/main/tools/pipeline.schema.json), [interface.schema.json](https://raw.githubusercontent.com/MaaXYZ/MaaFramework/refs/heads/main/tools/interface.schema.json).

--- a/README.md
+++ b/README.md
@@ -497,6 +497,33 @@ You can find an example project here: [Sandbox.SourceGenerator](./Solutions/Sand
 
 We'd like to credit our Google Summer of Code 2024 contributor, [Pranay Joshi](https://github.com/pranayjoshi) and mentor [Greg Dennis](https://github.com/gregsdennis) for their work on this tool.
 
+#### Configuring the source generator
+
+There are a number of global configuration options for the source generator. These can be added to a `PropertyGroup` in your `.csproj` file.
+
+e.g.
+
+```xml
+<PropertyGroup>
+   <CorvusJsonSchemaOptionalAsNullable>None</CorvusJsonSchemaOptionalAsNullable>
+</PropertyGroup>
+```
+
+`CorvusJsonSchemaOptionalAsNullable`
+  - `None` - Do not emit nullable properties for optional properties
+  - `NullOrUndefined` - Emit nullable properties for optional properties
+
+`CorvusJsonSchemaDisableOptionalNamingHeuristics`
+  - `False` - Enable optional naming heuristics [default]
+  - `True` - Disable optional naming heuristics
+
+`CorvusJsonSchemaDisabledNamingHeuristics`
+  - Semi-colon separated list of naming heuristics to disable. You can list the available name heuristics with the `generatejsonschematypes listNameHeuristics` command in the CLI.
+
+`CorvusJsonSchemaAlwaysAssertFormat`
+  - `False` - Respect the vocabulary's format assertion
+  - `True` - Always assert format assertions [default]
+
 ### New dynamic schema validation
 
 There is a new `Corvus.Json.Validator` assembly, containing a `JsonSchema` type.

--- a/Solutions/Corvus.Json.CodeGeneration.CSharp/Corvus.Json.CodeGeneration/CSharp/CodeGeneratorExtensions.Object.cs
+++ b/Solutions/Corvus.Json.CodeGeneration.CSharp/Corvus.Json.CodeGeneration/CSharp/CodeGeneratorExtensions.Object.cs
@@ -639,12 +639,12 @@ internal static partial class CodeGeneratorExtensions
                     }
 
                     string name = generator.GetMethodNameInScope("TryAsDependentSchemaFor", suffix: declaration.JsonPropertyName);
-                    string dotnetPropertyName =
-                        typeDeclaration
+                    PropertyDeclaration property = typeDeclaration
                             .PropertyDeclarations
-                            .Single(p => p.JsonPropertyName == declaration.JsonPropertyName)
-                            .DotnetPropertyName();
+                            .Single(p => p.JsonPropertyName == declaration.JsonPropertyName);
+                    string dotnetPropertyName = property.DotnetPropertyName();
                     string targetName = declaration.ReducedDepdendentSchemaType.FullyQualifiedDotnetTypeName();
+                    bool isNullable = typeDeclaration.OptionalAsNullable() && property.RequiredOrOptional == RequiredOrOptional.Optional;
 
                     generator
                         .AppendSeparatorLine()
@@ -661,15 +661,10 @@ internal static partial class CodeGeneratorExtensions
                         .AppendLineIndent("public bool ", name, "(out ", targetName, " result)")
                         .AppendLineIndent("{")
                         .PushIndent()
-                            .AppendLineIndent("if (this.", dotnetPropertyName, ".IsNotUndefined())")
-                            .AppendLineIndent("{")
-                            .PushIndent()
-                                .AppendLineIndent("result = this.As<", targetName, ">();")
-                                .AppendLineIndent("return result.IsValid();")
-                            .PopIndent()
-                            .AppendLineIndent("}")
+                            .AppendConditionalBackingValueCallbackIndent("Backing.JsonElement", "jsonElementBacking", (g, f) => AppendJsonBackedTest(g, typeDeclaration, f, property, targetName))
+                            .AppendConditionalBackingValueCallbackIndent("Backing.Object", "objectBacking", (g, f) => AppendObjectBackedTest(g, typeDeclaration, f, property, targetName))
                             .AppendSeparatorLine()
-                            .AppendLineIndent("result = ", targetName, ".Undefined;")
+                            .AppendLineIndent("result = default;")
                             .AppendLineIndent("return false;")
                         .PopIndent()
                         .AppendLineIndent("}");
@@ -678,6 +673,65 @@ internal static partial class CodeGeneratorExtensions
         }
 
         return generator;
+
+        static void AppendJsonBackedTest(CodeGenerator generator, TypeDeclaration typeDeclaration, string fieldName, PropertyDeclaration property, string targetName)
+        {
+            if (generator.IsCancellationRequested)
+            {
+                return;
+            }
+
+            generator
+                .AppendLineIndent("if (this.", fieldName, ".ValueKind != JsonValueKind.Object)")
+                .AppendLineIndent("{")
+                .PushIndent()
+                    .AppendLineIndent("result = default;")
+                    .AppendLineIndent("return false;")
+                .PopIndent()
+                .AppendLineIndent("}")
+                .AppendSeparatorLine()
+                .AppendLineIndent(
+                    "if (this.",
+                    fieldName,
+                    ".TryGetProperty(",
+                    generator.JsonPropertyNamesClassName(),
+                    ".",
+                    property.DotnetPropertyName(),
+                    "Utf8, out _))")
+                .AppendLineIndent("{")
+                .PushIndent();
+
+            generator
+                    .AppendLineIndent("result = this.As<", targetName, ">();")
+                    .AppendLineIndent("return result.IsValid();")
+                .PopIndent()
+                .AppendLineIndent("}");
+        }
+
+        static void AppendObjectBackedTest(CodeGenerator generator, TypeDeclaration typeDeclaration, string fieldName, PropertyDeclaration property, string targetName)
+        {
+            if (generator.IsCancellationRequested)
+            {
+                return;
+            }
+
+            generator
+                .AppendSeparatorLine()
+                .AppendLineIndent(
+                    "if (this.",
+                    fieldName,
+                    ".TryGetValue(",
+                    generator.JsonPropertyNamesClassName(),
+                    ".",
+                    property.DotnetPropertyName(),
+                    ", out _))")
+                .AppendLineIndent("{")
+                .PushIndent()
+                    .AppendLineIndent("result = this.As<", targetName, ">();")
+                    .AppendLineIndent("return result.IsValid();")
+                .PopIndent()
+                .AppendLineIndent("}");
+        }
     }
 
     /// <summary>

--- a/Solutions/Corvus.Json.SourceGenerator/Corvus.Json.SourceGenerator.csproj
+++ b/Solutions/Corvus.Json.SourceGenerator/Corvus.Json.SourceGenerator.csproj
@@ -156,6 +156,8 @@
     <None Include="$(PKGSystem_Text_Encoding_CodePages)\lib\netstandard2.0\System.Text.Encoding.CodePages.dll" Pack="true" PackagePath="analyzers/dotnet/cs" Visible="false" />
     <None Include="$(PKGSystem_Text_Encodings_Web)\lib\netstandard2.0\System.Text.Encodings.Web.dll" Pack="true" PackagePath="analyzers/dotnet/cs" Visible="false" />
     <None Include="$(PKGSystem_Threading_Tasks_Extensions)\lib\netstandard2.0\System.Threading.Tasks.Extensions.dll" Pack="true" PackagePath="analyzers/dotnet/cs" Visible="false" />
+    <!-- Package the props file -->
+    <None Include="Corvus.Json.SourceGenerator.props" Pack="true" PackagePath="build" Visible="false" />
   </ItemGroup>
 
   <ItemGroup>

--- a/Solutions/Corvus.Json.SourceGenerator/Metaschema.cs
+++ b/Solutions/Corvus.Json.SourceGenerator/Metaschema.cs
@@ -53,10 +53,10 @@ internal static class Metaschema
             "https://json-schema.org/draft/2019-09/meta/format",
             Parse(assembly, "metaschema.draft2019_09.meta.format.json"));
         documentResolver.AddDocument(
-            "https://json-schema.org/draft/2019-09/meta/hyper_schema",
+            "https://json-schema.org/draft/2019-09/meta/hyper-schema",
             Parse(assembly, "metaschema.draft2019_09.meta.hyper-schema.json"));
         documentResolver.AddDocument(
-            "https://json-schema.org/draft/2019-09/meta/meta_data",
+            "https://json-schema.org/draft/2019-09/meta/meta-data",
             Parse(assembly, "metaschema.draft2019_09.meta.meta-data.json"));
         documentResolver.AddDocument(
             "https://json-schema.org/draft/2019-09/meta/validation",
@@ -75,16 +75,16 @@ internal static class Metaschema
             "https://json-schema.org/draft/2020-12/meta/core",
             Parse(assembly, "metaschema.draft2020_12.meta.core.json"));
         documentResolver.AddDocument(
-            "https://json-schema.org/draft/2020-12/meta/format_annotation",
+            "https://json-schema.org/draft/2020-12/meta/format-annotation",
             Parse(assembly, "metaschema.draft2020_12.meta.format-annotation.json"));
         documentResolver.AddDocument(
-            "https://json-schema.org/draft/2020-12/meta/format_assertion",
+            "https://json-schema.org/draft/2020-12/meta/format-assertion",
             Parse(assembly, "metaschema.draft2020_12.meta.format-assertion.json"));
         documentResolver.AddDocument(
-            "https://json-schema.org/draft/2020-12/meta/hyper_schema",
+            "https://json-schema.org/draft/2020-12/meta/hyper-schema",
             Parse(assembly, "metaschema.draft2020_12.meta.hyper-schema.json"));
         documentResolver.AddDocument(
-            "https://json-schema.org/draft/2020-12/meta/meta_data",
+            "https://json-schema.org/draft/2020-12/meta/meta-data",
             Parse(assembly, "metaschema.draft2020_12.meta.meta-data.json"));
         documentResolver.AddDocument(
             "https://json-schema.org/draft/2020-12/meta/unevaluated",
@@ -94,10 +94,10 @@ internal static class Metaschema
             Parse(assembly, "metaschema.draft2020_12.meta.validation.json"));
 
         documentResolver.AddDocument(
-            "https://corvus-oss.org/json_schema/2020-12/schema",
+            "https://corvus-oss.org/json-schema/2020-12/schema",
             Parse(assembly, "metaschema.corvus.schema.json"));
         documentResolver.AddDocument(
-            "https://corvus-oss.org/json_schema/2020-12/meta/corvus_extensions",
+            "https://corvus-oss.org/json-schema/2020-12/meta/corvus-extensions",
             Parse(assembly, "metaschema.corvus.meta.corvus-extensions.json"));
 
         return documentResolver;

--- a/Solutions/Corvus.Json.Specs/Features/AdditionalSchema/Draft7/Repro492.feature
+++ b/Solutions/Corvus.Json.Specs/Features/AdditionalSchema/Draft7/Repro492.feature
@@ -1,0 +1,48 @@
+@draft7
+
+Feature: Repro 492 draft7
+
+Scenario Outline: Generation error with optional nullable and dependent schemas.
+	Given a schema file
+		"""
+		{
+			"$schema": "https://json-schema.org/draft-07/schema",
+			"type": "object",
+			"dependencies": {
+				"recognition": { }
+			}
+		}
+		"""
+	And the input data value <inputData>
+	And I generate a type for the schema with optional properties nullable
+	And I construct an instance of the schema type from the data
+	When I validate the instance
+	Then the result will be <valid>
+
+Examples:
+	| inputData                | valid |
+	| { "recognition": "foo" } | true  |
+
+
+Scenario Outline: Generation error with optional nullable and dependent schemas non-JsonAny
+	Given a schema file
+		"""
+		{
+			"$schema": "https://json-schema.org/draft-07/schema",
+			"type": "object",
+			"dependencies": {
+				"recognition": { "type": "object", "required": ["bar"], "properties": { "bar": {"type": "string" } } }
+			}
+		}
+		"""
+	And the input data value <inputData>
+	And I generate a type for the schema with optional properties nullable
+	And I construct an instance of the schema type from the data
+	When I validate the instance
+	Then the result will be <valid>
+
+Examples:
+	| inputData                                | valid |
+	| { "recognition": "foo", "bar": "there" } | true  |
+	| { "recognition": "foo" }                 | false |
+

--- a/Solutions/Sandbox.SourceGenerator/Sandbox.SourceGenerator.csproj
+++ b/Solutions/Sandbox.SourceGenerator/Sandbox.SourceGenerator.csproj
@@ -1,11 +1,15 @@
 ﻿<Project Sdk="Microsoft.NET.Sdk">
-
+  <Import Project="..\Corvus.Json.SourceGenerator\Corvus.Json.SourceGenerator.props"/>
   <PropertyGroup>
     <TargetFramework>net8.0</TargetFramework>
     <ImplicitUsings>enable</ImplicitUsings>
     <Nullable>enable</Nullable>
     <EmitCompilerGeneratedFiles>true</EmitCompilerGeneratedFiles>
     <IsPackable>false</IsPackable>
+  </PropertyGroup>
+
+  <PropertyGroup>
+    <CorvusJsonSchemaOptionalAsNullable>None</CorvusJsonSchemaOptionalAsNullable>
   </PropertyGroup>
 
   <ItemGroup>

--- a/Solutions/Sandbox.SourceGenerator/test.json
+++ b/Solutions/Sandbox.SourceGenerator/test.json
@@ -1,15 +1,23 @@
 {
     "$schema": "https://corvus-oss.org/json-schema/2020-12/schema",
-    "type": "array",
-    "prefixItems": [
-        {
-            "$corvusTypeName": "PositiveInt32",
-            "type": "integer",
-            "format": "int32",
-            "minimum": 0
-        },
-        { "type": "string" },
-        { "type":  "number", "minimum": 3.5 }
-    ],
-    "unevaluatedItems": false
+    "type": "object",
+    "properties": {
+        "theArray": {
+            "type": "array",
+            "prefixItems": [
+                {
+                    "$corvusTypeName": "PositiveInt32",
+                    "type": "integer",
+                    "format": "int32",
+                    "minimum": 0
+                },
+                { "type": "string" },
+                {
+                    "type": "number",
+                    "minimum": 3.5
+                }
+            ],
+            "unevaluatedItems": false
+        }
+    }
 }

